festim/reaction.py

```diff
@@ -48,61 +45,76 @@ class Reaction:
 
     def __init__(
         self,
-        reactant1: Union[F.Species, F.ImplicitSpecies],
-        reactant2: Union[F.Species, F.ImplicitSpecies],
-        product: F.Species,
+        reactant: Union[
+            F.Species, F.ImplicitSpecies, List[Union[F.Species, F.ImplicitSpecies]]
+        ],
+        product: Optional[Union[F.Species, List[F.Species]]],
         k_0: float,
         E_k: float,
         p_0: float,
         E_p: float,
         volume: F.VolumeSubdomain1D,
     ) -> None:
-        self.reactant1 = reactant1
-        self.reactant2 = reactant2
-        self.product = product
         self.k_0 = k_0
         self.E_k = E_k
-        self.p_0 = p_0
-        self.E_p = E_p
+        self.p_0 = p_0 or 0.0
+        self.E_p = E_p or 0.0
+        self.reactant = reactant
+        self.product = product or []
         self.volume = volume
 
     @property
-    def reactant1(self):
-        return self._reactant1
-
-    @reactant1.setter
-    def reactant1(self, value):
-        if not isinstance(value, (F.Species, F.ImplicitSpecies)):
-            raise TypeError(
-                f"reactant1 must be an F.Species or F.ImplicitSpecies, not {type(value)}"
+    def reactant(self):
+        return self._reactant
+
+    @reactant.setter
+    def reactant(self, value):
+        if not isinstance(value, list):
+            value = [value]
+        if len(value) == 0:
+            raise ValueError(
+                f"reactant must be an entry of one or more species objects, not an empty list."
             )
-        self._reactant1 = value
+        for i in value:
+            if not isinstance(i, (F.Species, F.ImplicitSpecies)):
+                raise TypeError(
+                    f"reactant must be an F.Species or F.ImplicitSpecies, not {type(i)}"
+                )
+        self._reactant = value
 
     @property
-    def reactant2(self):
-        return self._reactant2
-
-    @reactant2.setter
-    def reactant2(self, value):
-        if not isinstance(value, (F.Species, F.ImplicitSpecies)):
-            raise TypeError(
-                f"reactant2 must be an F.Species or F.ImplicitSpecies, not {type(value)}"
-            )
-        self._reactant2 = value
+    def product(self):
+        return self._product
+
+    @product.setter
+    def product(self, value):
+        if value == []:
+            if self.p_0 != 0.0:
+                raise ValueError(
+                    f"p_0 must be 0, not {self.p_0} when no products are present."
+                )
+            elif self.E_p != 0.0:
+                raise ValueError(
+                    f"E_p must be 0, not {self.E_p} when no products are present."
+                )
```

Comment:
This test should be here but in `reaction_term`.

Consider:

```python
my_reac = F.Reaction(reactant=A, product=[], p_0=1, E_p=1)  # raises an error
my_reac = F.Reaction(reactant=A, product=B, p_0=None, E_p=None) # doesn't raise an error
```

You would think you just have to make a setter for `p_0` and `E_p` but then you end up having a cyclic loop.

Instead, don't do this test in the setters but in `reaction_term()`

Reply:
Aha okay, good point. I was trying to avoid putting the test in `reaction_term()` as that is only called during the actual solve and not in instantiation of the class?

Reply:
The problem with doing it only at instantiation is that you can have it right on instantiation and then modify it afterwards (between instantiation and the call to reaction_term()) and break things

Reply:
Good call.